File: tree/rebuild_binarytree.py
#          A
#      B       C
#        D   E
#       F     G

# pre: A B D F C E G
#  in: B F D A E G C

from typing import List


class BinaryTree:

    def __init__(self, value: int) -> None:
        self.value = value
        self.lchild = None
        self.rchild = None


def recover_binarytree(preorder: List[str], inorder: List[str]) -> BinaryTree:
    if not (preorder and inorder):
        return

    return rebuild_binarytree(preorder, 0,
                              len(preorder) - 1, inorder, 0,
                              len(inorder) - 1)


def rebuild_binarytree(
    preorder: List[str],
    preorder_begin: int,
    preorder_end: int,
    inorder: List[str],
    inorder_begin: int,
    inorder_end: int,
) -> BinaryTree:
    if preorder_begin > preorder_end or inorder_begin > inorder_end:
        return None

    root_value = preorder[preorder_begin]
    root_node = BinaryTree(root_value)
    if preorder_begin == preorder_end or inorder_begin == inorder_end:
        return root_node

    root_inorder = inorder_begin
    while root_inorder <= inorder_end and inorder[root_inorder] != root_value:
        root_inorder += 1

    left_count = root_inorder - inorder_begin
    root_node.lchild = rebuild_binarytree(preorder, preorder_begin + 1,
                                          preorder_begin + left_count, inorder,
                                          inorder_begin, root_inorder - 1)

    root_node.rchild = rebuild_binarytree(preorder,
                                          preorder_begin + left_count + 1,
                                          preorder_end, inorder,
                                          root_inorder + 1, inorder_end)

    return root_node


if __name__ == "__main__":
    preorder = ['A', 'B', 'D', 'F', 'C', 'E', 'G']
    inorder = ['B', 'F', 'D', 'A', 'E', 'G', 'C']
    root = recover_binarytree(preorder, inorder)
    print(root)

File: tree/rebuild_binarytree.c
#include"stdio.h"
#include"stdlib.h"
#include"string.h"

// #          A
// #      B       C
// #        D   E
// #       F     G

// # pre: A B D F C E G
// #  in: B F D A E G C


typedef struct BinaryNode
{
    char value;
    struct BinaryNode* lchild, * rchild;
}*BinaryTree, BinaryNode;

BinaryTree rebuild_binarytree(char* pre, int pre_begin, int pre_end, char* in, int in_begin, int in_end) {
    if (pre_begin > pre_end || in_begin > in_end) {
        return NULL;
    }

    char root_value = pre[pre_begin];
    BinaryTree root = (BinaryTree)malloc(sizeof(BinaryNode));
    root->value = root_value;
    root->lchild = NULL;
    root->rchild = NULL;

    // 关键
    if (pre_begin == pre_end || in_begin == in_end) {
        return root;
    }

    int root_index = in_begin;

    // 关键
    while (root_index <= in_end && root_value != in[root_index])
    {
        root_index += 1;
    }

    int node_count = root_index - in_begin;
    root->lchild = rebuild_binarytree(pre, pre_begin + 1, pre_begin + node_count, in, in_begin, root_index - 1);
    root->rchild = rebuild_binarytree(pre, pre_begin + node_count + 1, pre_end, in, root_index + 1, in_end);

    return root;
}

BinaryTree recover_binarytree(char* pre, char* in) {
    if (pre == NULL || in == NULL) {
        return NULL;
    }

    return rebuild_binarytree(pre, 0, strlen(pre) - 1, in, 0, strlen(in) - 1);
}

// 前序遍历
void print_node(BinaryTree root) {
    if (root == NULL)
        return;

    printf("%c", root->value);
    print_node(root->lchild);
    print_node(root->rchild);

}


int main() {
    char pre[] = "ABDFCEG";
    char in[] = "BFDAEGC";
    BinaryTree root = recover_binarytree(pre, in);
    printf("%c, %c\n", root->lchild->value, root->rchild->value);
    print_node(root);
    printf("\n");
}